base `parseEnum` on a case statement, fixes #14030

Vindaar opened the pull request:
Changes the `parseEnum` implementation to be based on a case statement comparing the given normalized string value to the normalized field of the enum.

@cooldome: I'm not quite sure what you meant with:
```nim
nnkOfBranch.newTree(normalize(str), enumVal)
```
I don't understand why this should work, nor could I get it to work.

In any case, I had to change the implementation a little bit anyways. While writing a test, I put an enum into a `block`, which broke the code, since the fields weren't known anymore. So instead I determine the correct field value and use `T(fieldValue)` instead.

Vindaar committed:
base `parseEnum` on a case statement, fixes #14030

cooldome commented on lib/pure/strutils.nim:
```diff
@@ -1245,11 +1293,7 @@ proc parseEnum*[T: enum](s: string): T =
-    if cmpIgnoreStyle(s, $e) == 0:
-      return e
-  raise newException(ValueError, "invalid enum value: " & s)
+  genEnumStmt(T, s, default = false)
```
bool type is also enum, likely this example will trigger incorrect behavior:
```nim
parseEnum[bool]("xx", false) # exception instead of `false`
```
I suggest as default to use `nil` value

Vindaar replied:
Good catch, I forgot that that's how bools work.

edit: well, nevermind. Using `parseEnum` with `bool` results in a "cannot instantiate: `parseEnum[bool]`" error. Better to use `nil` anyways though.

cooldome commented on lib/pure/strutils.nim:
```diff
@@ -1229,6 +1230,53 @@ proc parseBool*(s: string): bool =
+        let enumFieldNum = f[1]
+        result.add addOfBranch(f[0].toStrLit, f[1], n, norm)
+        fieldNum = enumFieldNum.intVal
+      else: error("Invalid tuple syntax!")
```
```error("Invalid tuple syntax!", f[1])```

it will help with displaying lineinfo  next to error

cooldome commented on lib/pure/strutils.nim:
```diff
@@ -1229,6 +1230,53 @@ proc parseBool*(s: string): bool =
+        result.add addOfBranch(f[0].toStrLit, f[1], n, norm)
+        fieldNum = enumFieldNum.intVal
+      else: error("Invalid tuple syntax!")
+    else: error("Invalid node for enum type!")
```
error("Invalid node for enum type!", f)

cooldome commented:
Appologies in my example with `normalize` I didn't convert `NimNode` to string. I also realized that getImpl() gives you idents of enums, not symbols. IMO, it should be fixed in future.

This should work
```nim
proc addOfBranch(s: string, field, enumType: NimNode): NimNode =
  result = nnkOfBranch.newTree(newLit normalize(s),
    nnkDotExpr.newTree(enumType, field)) # `T.<fieldValue>`

macro genEnumStmt(typ, argSym: typed, default: typed): untyped =
  # generates a case stmt, which assigns the correct enum field given
  # a normalized string comparison to the `argIdent` input.
  # NOTE: for an enum with fields Foo, Bar, ... we cannot generate
  # `of "Foo".normalize: Foo`.
  # This will fail, if the enum is not defined at top level (e.g. in a block).
  # Thus we check for the field value of the (possible holed enum) and convert
  # the integer value to the generic argument `n`.
  let typ = typ.getTypeInst[1]
  let impl = typ.getImpl[2]
  expectKind impl, nnkEnumTy
  let norm = bindSym"normalize"
  result = nnkCaseStmt.newTree(nnkDotExpr.newTree(argSym, norm))
  for f in impl:
    case f.kind
    of nnkEmpty: continue # skip first node of `enumTy`
    of nnkSym, nnkIdent:
      result.add addOfBranch(f.strVal, f, typ)
    of nnkEnumFieldDef:
      case f[1].kind
      of nnkStrLit: 
        result.add addOfBranch(f[1].strVal, f[0], typ)
      of nnkTupleConstr:
        result.add addOfBranch(f[1][1].strVal, f[0], typ)
      of nnkIntLit:
        result.add addOfBranch(f[0].strVal, f[0], typ)
      else: error("Invalid tuple syntax", f[1])
    else: error("Invalid node for enum type", f)
  # finally add else branch to raise or use default
  if default == nil:
    let raiseStmt = quote do:
      raise newException(ValueError, "Invalid enum value: " & $`argSym`)
    result.add nnkElse.newTree(raiseStmt)
  else:
    expectKind(default, nnkSym)
    result.add nnkElse.newTree(default)

proc parseEnum*[T: enum](s: string): T =
  ## Parses an enum ``T``.
  ##
  genEnumStmt(T, s, default = nil)
```

Vindaar committed:
apply simplifactions / clean up, remove `norm` node, use strVal

cooldome commented on lib/pure/strutils.nim:
```diff
@@ -1229,6 +1230,54 @@ proc parseBool*(s: string): bool =
+    nnkCall.newTree(enumType, field) # `T(<fieldValue>)`
+  )
+
+macro genEnumStmt(typ, argSym: typed,
```
For better clarity we can more accurately specify arg type: `typ: typedesc`

cooldome commented on lib/pure/strutils.nim:
```diff
@@ -1229,6 +1230,54 @@ proc parseBool*(s: string): bool =
+proc addOfBranch(s: string, field, enumType: NimNode): NimNode =
+  result = nnkOfBranch.newTree(
+    newLit normalize(s),
+    nnkCall.newTree(enumType, field) # `T(<fieldValue>)`
```
With https://github.com/nim-lang/Nim/pull/14048 we can use field syminstead of `nnkCall.newTree(enumType, field)`

Vindaar replied:
Just tried it. Unfortunately that does not solve the problem if the enum is defined in a block. Still results in the same "undeclared identifer" error.

Vindaar committed:
export `normalize` in json.nim

Vindaar commented:
The test of `nimx` fails on MacOS, because of this enum here:

https://github.com/yglukhov/darwin/blob/master/darwin/foundation/nsnumber.nim#L8-L21

So this kind of means we cannot apply `normalize` to the argument and a field if the field has a string value attached to it, no?

edit: the old `parseEnum` implementation would have simply returned the first matching element (which may not be the correct one, if the argument was e.g. `"C"`! So parsing such enums was definitely broken with the old `parseEnum`). How should we handle this? @yglukhov, what do you say?

cooldome commented:
I think it is a bug in `normalize`. The first character of identifier is case sensitive while rest are not.

Vindaar commented:
The doc string for `normalize` explicitly states though:

```
That means to convert it to lower case and remove any '_'. This
should NOT be used to normalize Nim identifier names.
```

But still, I'm not sure if it's reasonable to argue that capitalization should only make sense for the first character of the string values? For the actual enum field names, sure. 

There don't seem to be well defined rules how string values are treated, which makes me believe that they are just considered as they are?

And well, apparently even this is allowed:
```nim
type
  Enum = enum
    A = "A"
    B = "A"
    C = "A"

let a = A
let b = B
let c = C
```

cooldome commented:
IMO,
```nim
type
  Enum = enum
    A = "A"
    B = "A"
    C = "A"
```
should give a reasonable compile time error user can understand. You can't parse this one.

For the rest we should use standard nim identifier convention:
```nim
proc nimIdentNormalize*(s: string): string =
  ## Normalizes the string `s`as Nim identifier.
  ##
  ## That means to convert to lower case and remove any '_' on all characters except
  ## first one.
  runnableExamples:
    doAssert normalize("Foo_bar") == "Foobar"
  result = newString(s.len)
  if s.len > 0:
    result[0] = s[0]
  var j = 0
  for i in 1..len(s) - 1:
    if s[i] in {'A'..'Z'}:
      result[j] = chr(ord(s[i]) + (ord('a') - ord('A')))
      inc j
    elif s[i] != '_':
      result[j] = s[i]
      inc j
  if j != s.len: setLen(result, j)
```

Vindaar committed:
cmp using nimIdentNormalize, error at CT if ambiguous enum found

`nimIdentNormalize` provided by @cooldome.

We track all names of the branches we have created so far and error if
a duplicate is found.

Dummy change to make github react...

Vindaar commented:
Ah, thanks to GitHub being messed up again, my commit didn't appear.

I used your `nimIdentNormalize` impl @cooldome (well, fixed it `j = 1` first) and now track which field names have already been encountered. If duplicates are found a CT error is thrown.

cooldome commented on lib/pure/strutils.nim:
```diff
@@ -1229,8 +1250,65 @@ proc parseBool*(s: string): bool =
+        fNum = f[1].intVal
+      else: error("Invalid tuple syntax!")
+    else: error("Invalid node for enum type!")
+    # add field if string not already added
```
Once call to nimIdentNormalize(fStr) can be saved:
```fStr = nimIdentNormalize```
and then use `fStr`

Vindaar committed:
fix docstring of `nimIdentNormalize`

cooldome commented:
Looks good, I made a couple of minor comments on style in code review. Please have a look.
Otherwise very good work

Vindaar commented:
I think your comments may have fallen victim to GitHub's outage. Can't see them.

cooldome commented:
I hope you see my comments now

Vindaar committed:
make `typ` arg `typedesc`, add lineinfo, call norm. only once

Vindaar commented:
Yes, I could see your comments now. I rebased on current devel and applied the changes, except the one to use the symbol directly. That still doesn't work if the enum isn't defined at top level.

cooldome commented:
Test failures are unrelated, merging
